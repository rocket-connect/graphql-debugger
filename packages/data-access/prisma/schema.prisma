generator client {
  provider        = "prisma-client-js"
  output          = "../node_modules/.prisma/client"
  previewFeatures = ["tracing"]
}

datasource db {
  provider = "sqlite"
  url      = "file:./dev.db"
}

model TraceGroup {
  id        String   @id @default(uuid())
  createdAt DateTime @default(now())
  updatedAt DateTime @updatedAt
  traceId   String   @unique
  spans     Span[]   @relation()
  schema    Schema?  @relation(fields: [schemaId], references: [id])
  schemaId  String?
}

model Span {
  id                String      @id @default(uuid())
  spanId            String
  parentSpanId      String?
  traceId           String
  name              String
  kind              String
  startTimeUnixNano BigInt
  endTimeUnixNano   BigInt
  durationNano      BigInt
  isForeign         Boolean     @default(true)
  isGraphQLRootSpan Boolean     @default(false)
  graphqlDocument   String?
  graphqlVariables  String?
  graphqlResult     String?
  graphqlContext    String?
  graphqlSchemaHash String?
  attributes        String?
  errorMessage      String?
  errorStack        String?
  createdAt         DateTime    @default(now())
  updatedAt         DateTime    @updatedAt
  traceGroup        TraceGroup? @relation(fields: [traceGroupId], references: [id])
  traceGroupId      String?
}

model Schema {
  id          String       @id @default(uuid())
  createdAt   DateTime     @default(now())
  updatedAt   DateTime     @updatedAt
  hash        String       @unique
  name        String?
  typeDefs    String
  traceGroups TraceGroup[]
}
