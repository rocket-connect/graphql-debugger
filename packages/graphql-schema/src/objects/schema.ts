import { prisma } from "@graphql-debugger/data-access";
import { ObjectRef } from "@graphql-debugger/utils";
import { builder } from "../schema";
import { TraceObject } from "./trace";
import { Schema } from "@graphql-debugger/types";

export const SchemaObject: ObjectRef<Schema> = builder.objectType("Schema", {
  fields: (t) => ({
    id: t.exposeID("id"),
    hash: t.exposeString("hash"),
    name: t.exposeString("name", { nullable: true }),
    typeDefs: t.exposeString("typeDefs"),
    createdAt: t.exposeString("createdAt"),
    traceGroups: t.field({
      type: [TraceObject],
      resolve: async (root) => {
        const traceGroups = await prisma.traceGroup.findMany({
          where: {
            schemaId: root.id,
          },
        });

        return traceGroups.map((traceGroup) => ({
          id: traceGroup.id,
          traceId: traceGroup.traceId,
          spans: [],
        }));
      },
    }),
  }),
});
