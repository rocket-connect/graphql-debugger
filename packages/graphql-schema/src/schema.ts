import { traceSchema } from "@graphql-debugger/trace-schema";
import { SchemaBuilder } from "@graphql-debugger/utils";
import { Objects } from "./objects";
import { Context } from "./context";
import { TRACE_SCHEMA } from "./config";

export const builder = new SchemaBuilder<{
  Objects: Objects;
  Context: Context;
}>({});

builder.queryType({});
builder.mutationType({});

require("./queries");
require("./mutations");

const build = builder.toSchema();

export const schema = TRACE_SCHEMA ? traceSchema({ schema: build }) : build;
