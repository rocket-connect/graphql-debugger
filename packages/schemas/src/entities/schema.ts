import { z } from "zod";
import { TraceSchema } from "./trace";

export const SchemaSchema = z.object({
  id: z.string(),
  hash: z.string(),
  name: z.string().optional().nullable(),
  typeDefs: z.string(),
  traceGroups: z.array(TraceSchema),
  createdAt: z.string(),
});
