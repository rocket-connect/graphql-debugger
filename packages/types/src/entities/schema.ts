import { SchemaSchema, z } from "@graphql-debugger/schemas";

export type Schema = z.infer<typeof SchemaSchema>;
