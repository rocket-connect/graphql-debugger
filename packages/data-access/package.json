{
  "name": "@graphql-debugger/data-access",
  "version": "0.0.0-alpha.48",
  "main": "build/index.js",
  "types": "build/index.d.ts",
  "files": [
    "build",
    "prisma"
  ],
  "license": "MIT",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/rocket-connect/graphql-debugger.git"
  },
  "homepage": "https://github.com/rocket-connect/graphql-debugger#readme",
  "scripts": {
    "build": "tsc",
    "postinstall": "npx prisma generate && npx prisma db push"
  },
  "peerDependencies": {
    "graphql": "^16.0.0 || ^17.0.0"
  },
  "dependencies": {
    "@graphql-debugger/utils": "workspace:^",
    "prisma": "5.2.0",
    "@prisma/client": "5.2.0"
  },
  "devDependencies": {
    "@faker-js/faker": "8.0.2",
    "gql-tag": "^1.0.1",
    "@graphql-tools/schema": "10.0.0",
    "@graphql-tools/utils": "10.0.6"
  }
}
