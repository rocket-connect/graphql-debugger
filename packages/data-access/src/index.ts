import { PrismaInstrumentation } from "@prisma/instrumentation";

import { PrismaClient } from ".prisma/client";

export const prisma = new PrismaClient();

export const tracing = new PrismaInstrumentation({
  middleware: false,
  ...(process.env.TRACE_PRISMA && {
    enabled: true,
  }),
});

if (!process.env.TRACE_PRISMA) {
  tracing.disable();
}

export * from ".prisma/client";
export * from "./clear-db";
