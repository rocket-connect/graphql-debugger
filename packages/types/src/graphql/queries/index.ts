export * from "./list-schemas";
export * from "./list-trace-groups";
export * from "./aggregate-spans";
