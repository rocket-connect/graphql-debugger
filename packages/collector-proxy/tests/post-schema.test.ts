import { prisma } from "@graphql-debugger/data-access";
import { hashSchema } from "@graphql-debugger/utils";

import { faker } from "@faker-js/faker";
import { makeExecutableSchema } from "@graphql-tools/schema";
import { describe, expect, test } from "@jest/globals";
import { parse, print } from "graphql";

import { request } from "./utils";

describe("POST /v1/schema", () => {
  test("should throw when no body is sent", async () => {
    const response = await request().post("/v1/schema").send({});

    expect(response.status).toBe(400);

    const body = await response.body;

    expect(body.message).toMatchSnapshot();
  });

  test("should throw an error when parsing the schema", async () => {
    const response = await request().post("/v1/schema").send({
      schema: "invalid schema",
    });

    expect(response.status).toBe(400);

    const body = await response.body;

    expect(body.message).toEqual("Could not parse schema");
  });

  test("shold return 200 when the schema is already stored", async () => {
    const schema = `
      type Query {
        name: String!
      }
    `;

    const executableSchema = makeExecutableSchema({
      typeDefs: schema,
      noLocation: true,
    });

    const hash = hashSchema(executableSchema);

    await prisma.schema.create({
      data: {
        hash,
        typeDefs: print(parse(schema)),
      },
    });

    const response = await request().post("/v1/schema").send({
      schema,
    });

    expect(response.status).toBe(200);
  });

  test("should store the schema", async () => {
    const schema = `
      type Query {
        ${faker.string.alpha({ length: 8 })}: String!
      }
    `;

    const response = await request().post("/v1/schema").send({
      schema,
    });

    expect(response.status).toBe(200);

    const hash = hashSchema(
      makeExecutableSchema({
        typeDefs: schema,
        noLocation: true,
      }),
    );

    const foundSchema = await prisma.schema.findFirst({
      where: {
        hash,
      },
    });

    expect(foundSchema).not.toBeNull();

    expect(foundSchema?.typeDefs).toEqual(print(parse(schema)));
  });
});
