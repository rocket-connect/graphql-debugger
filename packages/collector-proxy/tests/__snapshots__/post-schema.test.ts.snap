// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`POST /v1/schema should throw when no body is sent 1`] = `
"[
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "undefined",
    "path": [
      "body",
      "schema"
    ],
    "message": "Required"
  }
]"
`;
